### DISCUSSION TITLE
Improve getAllAppsWindowsTitle()

### DISCUSSION THREAD

**Discussion by zak-45** [Kalmat/PyWinCtl#93] [Q&A] [answered]
Hi, for one of my project, I need to retreive hWnd. 

I want to show to the end user a window pop up with all active windows grouped by apps. 
getAllAppsWindowsTitle is perfect for that, but missing the hWnd information.

I create this small script for this purpose :
```
import pywinctl as pwc
windows=pwc.getAllWindows()
for win in windows:
    print(win.title, win._hWnd,win.getAppName())

```

this work, except that by adding `win.getAppName()` this take too much time.

Have tried several solution, no one quickest.

Do you have any idea on how to improve that ?
And as suggestion: including hWnd into getAllAppsWindowsTitle() or create getAllAppsWindowsTitlehWnd() will be great.

and thanks for this great module !!

**Comment by Kalmat** [comment]
Hi! Thank you for your interest in this module!

First off, sorry for my late reply. I was trying to figure out how to properly manage your request. To be honest, I don't like mixing titles and names together with id's... On the other hand, you're totally right. `getAppName()` uses a for, so it is very slow when using it inside another for.

Still not sure what could be the best for all cases, but right now I'm considering to build a method which returns a dictionary including all windows info (app name and pid, window id, title, position, etc.)... If it is not too demanding (this is what I have to check in all three platforms), may be this a more general solution for other future use cases including yours.

Please, let me think and test the best solution. If you have any idea, it's welcome!

**Comment by Kalmat** [comment]
Sorry, I tested your code today... the time it takes is not acceptable at all!!!

I will work in a new version. I think, by now, that the best solution is to build a dict containing all info.

In the mean time, please use this code as a workaround (you are on Windows, right? Or do you need a multi-platform solution? If so, please let me know):

```
import win32process

import pywinctl as pwc
from pywinctl._pywinctl_win import _getAllApps


def getAppWindowsID(tryToFilter: bool = True):
    process_list = _getAllApps(tryToFilter)
    result = []
    for win in pwc.getAllWindows():
        id = win.getHandle()
        pID = win32process.GetWindowThreadProcessId(id)
        for item in process_list:
            appPID, appName = item
            if appPID == pID[1]:
                result.append([appName, win.title, id])
                break
    return result
```

It returns a list of other lists each containing app name, window title and window id. It takes around 0.2 seconds in my system (the dict will take a bit longer, but it will contain more info).

Note that `tryToFilter` will try to get user (non-system) applications only if you set it to `True`, but some real user applications are skipped in some cases (I didn't find a way to trustfully return only user applications in windows). So set it to `True` or `False` at your convenience.

Thank you again!

**Reply by zak-45** [reply]
thanks to you and not worry about time, we all have some other things to do.... In fact, yes, goal is to have a cross-platform solution, so a dict with all informations is ok for me. My project is still on beta phase, so do your best at your time. I just wanted to point this sad effect by using `getAppName()`.

**Comment by Kalmat** [comment, marked as answer]
Hi! I have uploaded a new version to PyPi (v0.4.01) which includes `getAllWindowsDict()` function.

Whenever you have time, please check it and let me know if it works in your case.

Thanks for pointing out this issue and your help!

**Comment by zak-45** [comment]
Hi, this works, great job, thanks  a lot. 

Just one remark: to be "Json friendly", value for 'Point' and 'size' need to be reviewed (put ' or made it serializable).
For reference, here is my working code:

`    @staticmethod
    async def windows_titles():
        """ Provide a list of all window titles / hWnd by applications """

        try:
            # Assuming these are your custom classes
            class Point:
                def __init__(self, x, y):
                    self.x = x
                    self.y = y

            class Size:
                def __init__(self, width, height):
                    self.width = width
                    self.height = height

            # Define a function to serialize custom objects
            def custom_serializer(obj):
                if isinstance(obj, Point):
                    return {"x": obj.x, "y": obj.y}
                elif isinstance(obj, Size):
                    return {"width": obj.width, "height": obj.height}
                raise TypeError(f"Type {type(obj)} not serializable")

            # Your dictionary
            data = pwc.getAllWindowsDict()
            # Convert dictionary to JSON
            windows_all = json.dumps(data, default=custom_serializer, ensure_ascii=False, indent=4)
            windows_by_app = json.loads(windows_all)

        except Exception as er:
            logger.error(f"Error retrieving windows: {er}")
            return {}

        return windows_by_app
`

I will close this one as complete and thanks again.

**Comment by Kalmat** [comment]
So happy to hear that!

Regarding Point and Size objects, you're right... Inside a dictionary perhaps it's better to stick to standard types (tuple[int, int]). I will change that in next version.

In the mean time, just for your info, they are a NamedTuple classes:

```
class Point(NamedTuple):
    """Container class to handle Point struct (x, y)"""
    x: int
    y: int

class Size(NamedTuple):
    """Container class to handle Size struct (width, height)"""
    width: int
    height: int
```

So you can use them just importing them from PyWinCtl:

`
from pywinbox import Point, Size
`

You can also serialize them in a more simple way by using this:

```
if isinstance(obj, Point) or isinstance(obj, Size):
    return tuple(obj)
```

Thank you!
